Userland/Applications/PixelPaint/VectorscopeWidget.cpp

```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    return {};
+}
+
+void VectorscopeWidget::rebuild_vectorscope_image()
+{
+    m_vectorscope_image = MUST(Gfx::Bitmap::try_create(Gfx::BitmapFormat::BGRA8888, size()));
+    m_vectorscope_image->fill(Color::Transparent);
+
+    Gfx::Painter base_painter(*m_vectorscope_image);
+    Gfx::AntiAliasingPainter painter(base_painter);
+
+    auto const scope_size = min(height(), width());
+    auto const min_scope_size = parent_widget()->min_height().as_int();
+    auto const color_vector_scale = scope_size / static_cast<float>(min_scope_size);
+    auto const size_1x1 = Gfx::FloatSize { 2.5f, 2.5f } * static_cast<float>(color_vector_scale);
+
+    base_painter.translate(width() / 2, height() / 2);
+    painter.translate(static_cast<float>(width()) / 2.0f, static_cast<float>(height()) / 2.0f);
+    for (size_t u_index = 0; u_index < u_v_steps; ++u_index) {
+        for (size_t v_index = 0; v_index < u_v_steps; ++v_index) {
+            auto const color_vector = ColorVector::from_indices(u_index, v_index);
+            auto const brightness = m_vectorscope_data[u_index][v_index];
+            if (brightness < 0.0001)
+                continue;
+            auto const pseudo_rect = Gfx::FloatRect::centered_at(color_vector.to_vector(scope_size) * 2.0, size_1x1);
+            auto color = Color::from_yuv(0.6, color_vector.u, color_vector.v);
+            color.saturate_to(1.0 - min(brightness, 1.0));
+            color.set_alpha(static_cast<u8>(min(sqrt(brightness), alpha_range) * NumericLimits<u8>::max() / alpha_range));
+            painter.fill_rect(pseudo_rect, color);
+        }
+    }
+}
+
+void VectorscopeWidget::paint_event(GUI::PaintEvent& event)
+{
+    GUI::Painter base_painter(*this);
+    Gfx::AntiAliasingPainter painter(base_painter);
+    base_painter.add_clip_rect(event.rect());
+    // From this point on we're working with 0,0 as the scope center to make things easier.
+    base_painter.translate(width() / 2, height() / 2);
+    painter.translate(static_cast<float>(width()) / 2.0f, static_cast<float>(height()) / 2.0f);
+
+    auto const graticule_color = Color::White;
+    auto const scope_size = min(height(), width());
+    auto const graticule_size = scope_size / 6;
+    auto const graticule_thickness = graticule_size / 12;
+    auto const entire_scope_rect = Gfx::FloatRect::centered_at({ 0, 0 }, { scope_size, scope_size });
+
+    painter.fill_ellipse(entire_scope_rect.to_rounded<int>().shrunken(graticule_thickness * 2, graticule_thickness * 2), Color::Black);
+
+    // Main scope data
+    if (m_image) {
+        if (m_vectorscope_image->size() != this->size())
+            rebuild_vectorscope_image();
+
+        base_painter.blit({ -width() / 2, -height() / 2 }, *m_vectorscope_image, m_vectorscope_image->rect());
+    }
+
+    // Graticule(s)
+    painter.draw_ellipse(entire_scope_rect.to_rounded<int>(), graticule_color, graticule_thickness);
+
+    // FIXME: Translation is apparently messed up with line drawing.
```

Comment:
Needs clarification to be actionable

Reply:
Well, the translation calls to the painters don't appear to work correctly, and I figured out a combination of calls through trial and error that do what I want, but I don't know how they do that. Translation does work correctly with things like rectangle and text drawing for example, so it's very strange. I don't know what I should clarify here.

Reply:
Put that as a comment :)
Just think if someone not familiar with the problem could get started on a fix with the current comment vs what you wrote now.